package-tests.cake

```diff
@@ -0,0 +1,312 @@
+            {
+                Total = 2 * 37,
+                Passed = 2 * 23,
+                Failed = 2 * 5,
+                Warnings = 0,
+                Inconclusive = 2 * 1,
+                Skipped = 2 * 7
+            }));
+    }
+}
+
+public class NuGetNetFXPackageTester : NetFXPackageTester
+{
+    public NuGetNetFXPackageTester(ICakeContext context, string packageVersion)
+        : base(context, packageVersion) { }
+
+    protected override string PackageName => $"NUnit.ConsoleRunner.{_packageVersion}.nupkg";
+    protected override string PackageInstallDirectory => _packageDir + "test/nuget-netfx/";
+    protected override string PackageBinDir => PackageInstallDirectory + "tools/";
+}
+
+public class NuGetNetCorePackageTester : NetCorePackageTester
+{
+    public NuGetNetCorePackageTester(ICakeContext context, string packageVersion)
+        : base(context, packageVersion) { }
+
+    protected override string PackageName => $"NUnit.ConsoleRunner.NetCore.{_packageVersion}.nupkg";
+    protected override string PackageInstallDirectory => _packageDir + "test/nuget-netcore/";
+    protected override string PackageBinDir => PackageInstallDirectory + "tools/netcoreapp3.1/any/";
+}
+
+public class ChocolateyPackageTester : NetFXPackageTester
+{
+    public ChocolateyPackageTester(ICakeContext context, string packageVersion)
+        : base(context, packageVersion) { }
+
+    protected override string PackageName => $"nunit-console-runner.{_packageVersion}.nupkg";
+    protected override string PackageInstallDirectory => _packageDir + "test/choco/";
+    protected override string PackageBinDir => PackageInstallDirectory + "tools/";
+}
+
+public class MsiPackageTester : NetFXPackageTester
+{
+    public MsiPackageTester(ICakeContext context, string packageVersion)
+        : base(context, packageVersion)
+    {
+        // Add tests specific to the msi package
+        PackageTests.Add(new PackageTest(
+            "Run project with both copies of mock-assembly",
+            $"../../NetFXTests.nunit --config={_config}",
+            new ExpectedResult("Failed")
+            {
+                Total = 2 * 37,
+                Passed = 2 * 23,
+                Failed = 2 * 5,
+                Warnings = 0,
+                Inconclusive = 2 * 1,
+                Skipped = 2 * 7
+            }));
+    }
+
+    protected override string PackageName => $"NUnit.Console-{_packageVersion}.msi";
+    protected override string PackageInstallDirectory => _packageDir + "test/msi/";
+    protected override string PackageBinDir => PackageInstallDirectory + "NUnit.org/nunit-console/";
+
+    protected override void CreatePackageInstallDirectory()
+    {
+        // Msiexec does not tolerate forward slashes!
+        string package = PackageUnderTest.ToString().Replace("/", "\\");
+        string testDir = PackageInstallDirectory.Replace("/", "\\");
+        int rc = _context.StartProcess("msiexec", $"/a {package} TARGETDIR={testDir} /q");
```

Comment:
I'm not sure - but I wonder if this will throw and terminate the script if msiexec cannot be found? Should we instead try/catch, show an error/fail the tests but allow the other tests to continue running?

I checked the Cake docs to try and find out, but it's not documented unfortunately.

Reply:
Not sure about Windows - but presumably it wouldn't be on Linux/OSX? Seems sensible to me. 

When I tried to run the previous PR, I thought missing msiexec was the reason it failed, but I think it was missing wix instead. Haven't been able to pull this one yet to recheck, I'm afraid!

Reply:
Actually, I guess choco is Windows-only too, right? Maybe we should just add IsRunningOnWindows to these two tasks, so people could still theoretically run the nuget tests on Linux, if required.

Reply:
I had assumed we were running the packaging on Linux, but I just checked and see we aren't. I'll run some tests locally to see if it builds and if not I'll prevent the msi and chocolatey stuff from executing.

I had assumed we were testing this stuff under Azure but it appears that we use the Package target only on the windows build. Since I'm not an Azure guy, I think any fix to that (if you want one) should be separate.